mllib/src/main/scala/org/apache/spark/ml/tree/impl/RandomForest.scala

```diff
@@ -122,8 +132,9 @@ private[spark] object RandomForest extends Logging with Serializable {
     timer.start("init")
 
     val retaggedInput = input.retag(classOf[Instance])
-    val metadata =
+    val metadata = precomputedMetadata.getOrElse {
```

Comment:
Is this the same for all trees -- it's a property of all of the input?
Also are there callers that can't precompute and pass this -- do you need Option?

Reply:
Yes
the input `RDD[Instance]` for each tree are differect (because labels are different), but the computation of `DecisionTreemetadata` do use labels.
So it is the same for all trees.

Reply:
> are there callers that can't precompute and pass this

this method is indirectly used in DecisionTree/RandomForest in both .ml/.mllib sides.
Other algs will compute `DecisionTreemetadata` only once, in this method.

It will cause many changes if we precomupte `DecisionTreemetadata` outside and pass it to this method.